### PR TITLE
feat(share/eds): Store nmt nodes on failed reconstruction

### PR DESCRIPTION
**Pull request by walldiss**
Store repaired shares and calculated nmt proofs if reconstruction failed.

### CODE REVIEW

**Review by Wondertan — DISMISSED** [review]
Looks correct. I feel like this needs a test, which will stay for after shwap times.

**Review by Wondertan — DISMISSED** [review]

File: share/eds/retriever.go [hunk 1]
```diff
@@ -123,13 +128,24 @@ type retrievalSession struct {
...
-		visitor := ipld.ProofsAdderFromCtx(ctx).VisitFn()
-		if visitor != nil {
-			opts = append(opts, nmt.NodeVisitor(visitor))
+
+		proofsAdder := ipld.ProofsAdderFromCtx(ctx)
+		visitor := func(hash []byte, children ...[]byte) {
+			if proofsAdder != nil {
+				proofsAdder.VisitFn()(hash, children...)
+			}
+			adder.Visit(hash, children...)
```
Comment (renaynay) [inline share/eds/retriever.go:146]: so in case of proofsAdder in ctx, both retrieval session's adder + proofs adder from ctx will cache collected proofs ?

**Review by renaynay — APPROVED** [review]

File: share/eds/retriever.go (hunk 1 above)
Comment (walldiss) [reply share/eds/retriever.go:146]: yes. Not a big deal, because soon, when shwap is released, we would not need proofsAdder here

**Review by renaynay — APPROVED** [review]
Agree with hlib that this needs a test case -- I think we should make an issue or document some test cases somewhere that we need to implement.

**Comment by walldiss** [comment]
> Agree with hlib that this needs a test case -- I think we should make an issue or document some test cases somewhere that we need to implement.

https://github.com/celestiaorg/celestia-node/issues/3314

**Review by Wondertan — APPROVED** [review]

**Review by vgonkivs — APPROVED** [review]

**Pull request merged**